basalt/src/config/mod.rs

```diff
@@ -0,0 +1,106 @@
+mod keybinding;
+#[cfg(test)]
+mod test;
+
+use std::{collections::HashMap, fs::read_to_string, path::PathBuf, str::from_utf8};
+
+use etcetera::{choose_base_strategy, home_dir, BaseStrategy};
+use keybinding::KeyBinding;
+use serde::Deserialize;
+
+use crate::app::Action;
+
+pub(crate) use keybinding::Key;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Config {
+    pub keymap: HashMap<Key, Action>,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        let TomlConfig { key_bindings } = TomlConfig::default();
+
+        Self {
+            keymap: key_bindings
+                .expect("The built-in config.toml should contain key bindings")
+                .into_iter()
+                .map(|KeyBinding { key, command }| (key, command.into()))
+                .collect(),
+        }
+    }
+}
+
+impl From<TomlConfig> for Config {
+    fn from(TomlConfig { key_bindings }: TomlConfig) -> Self {
+        let mut config = Self::default();
+        config.merge_key_bindings(key_bindings);
+
+        config
+    }
+}
+
+impl Config {
+    pub fn build() -> Self {
+        TomlConfig::build().map_or_else(Self::default, Self::from)
+    }
+    fn merge_key_bindings(&mut self, key_bindings: Option<Vec<KeyBinding>>) {
+        if let Some(key_bindings) = key_bindings {
+            key_bindings
+                .into_iter()
+                .for_each(|KeyBinding { key, command }| {
+                    let new_action: Action = command.into();
+                    self.keymap
+                        .entry(key)
+                        .and_modify(|old_action| *old_action = new_action)
+                        .or_insert(new_action);
+                })
+        }
+    }
```

Comment:
I would change this to be merging complete configs instead! And to retain immutability with this function, we could pass the defined key_bindings in the configuration as the starting point and initialize a fold with that, however, this is not as performant as we are doing memory allocations.

```suggestion
    /// Takes self and another config and merges the `key_bindings` together overwriting the
    /// existing entries with the value from another config.
    pub(crate) fn merge(&self, config: Config) -> Config {
        config
            .key_bindings
            .into_iter()
            .fold(self.key_bindings.clone(), |mut acc, (key, value)| {
                acc.entry(key)
                    .and_modify(|v| *v = value.clone())
                    .or_insert(value);
                acc
            })
            .into()
    }
```

Reply:
I thought I'd make a function for each section of the configuration.
`config.merge_key_bindings(key_bindings);` ( in `From<TomlConfig> for Config` ) was to be the first.